### PR TITLE
Release

### PR DESCRIPTION
**Pull request by MulyukovAidar**

### CODE REVIEW

File: src/sankeydiagram.ts [hunk 1]
```diff
@@ -25,18 +25,22 @@
 */
 import "../style/visual.less";
 import powerbi from "powerbi-visuals-api";
-import "@babel/polyfill";
+import "core-js/stable";
```
Comment (uve) [inline src/sankeydiagram.ts]: Move core-js initialization to constructor method like here:
https://github.com/microsoft/powerbi-visuals-gantt/blob/master/src/gantt.ts#L378-L380

File: src/sankeydiagram.ts [hunk 2]
```diff
@@ -368,6 +416,17 @@ export class SankeyDiagram implements IVisual {
         this.main.attr("transform", translate(this.margin.left, this.margin.top));
     }
 
+    private saveLinkProperties(link, foundLink) {
```
Comment (uve) [inline src/sankeydiagram.ts]: Add interfaces to **link** and **fountLink** arguments

File: src/sankeydiagram.ts [hunk 3]
```diff
@@ -420,20 +475,46 @@ export class SankeyDiagram implements IVisual {
...
+        if (settings.filterDetected) {
+            settings._linkPositions.forEach(l => l.display = true);
+            if (settings.general.pinNodes) {
+                settings._linkPositions.forEach(link => {
+                    link.identity = this.visualHost.createSelectionIdBuilder().createSelectionId();
+                    link.display = false;
+                });
+                links.forEach(link => {
+                    let foundLink: SankeyDiagramLink = settings._linkPositions.find(slink => slink.source.label.name === link.source.label.name && slink.destination.label.name === link.destination.label.name);
+                    if (foundLink) { this.saveLinkProperties(link, foundLink); }
```
Comment (uve) [inline src/sankeydiagram.ts]: do not use one-liner here

File: src/sankeydiagram.ts [hunk 4]
```diff
@@ -420,20 +475,46 @@ export class SankeyDiagram implements IVisual {
...
+        settings.filterDetected = savedNodesArray.length !== nodes.length;
+        if (settings.filterDetected) {
+            settings._linkPositions.forEach(l => l.display = true);
+            if (settings.general.pinNodes) {
+                settings._linkPositions.forEach(link => {
+                    link.identity = this.visualHost.createSelectionIdBuilder().createSelectionId();
+                    link.display = false;
+                });
+                links.forEach(link => {
+                    let foundLink: SankeyDiagramLink = settings._linkPositions.find(slink => slink.source.label.name === link.source.label.name && slink.destination.label.name === link.destination.label.name);
```
Comment (uve) [inline src/sankeydiagram.ts]: replace the following condition by calling one of two possible methods:
- to compare two values: **source** and **destinations**, and cover it by unit-tests.
- to search all links by your link.

slink.source.label.name === link.source.label.name && slink.destination.label.name === link.destination.label.name

File: src/sankeydiagram.ts [hunk 5]
```diff
@@ -420,20 +475,46 @@ export class SankeyDiagram implements IVisual {
...
+        }
+        else {
+            links.forEach(link => {
+                let foundLink: SankeyDiagramLink = settings._linkPositions.find(slink => slink.source.label.name === link.source.label.name && slink.destination.label.name === link.destination.label.name);
+                if (foundLink) {
+                    this.saveLinkProperties(link, foundLink);
+                }
+            });
+            sankeyDiagramDataView.nodes = savedNodesArray;
+            sankeyDiagramDataView.links = settings._linkPositions;
```
Comment (uve) [inline src/sankeydiagram.ts]: Are **links** and **_linkPositions** variables the same? If so, why do you use different names?

File: src/sankeydiagram.ts [hunk 6]
```diff
@@ -548,17 +625,39 @@ export class SankeyDiagram implements IVisual {
     }
 
     private restoreNodePositions(nodes: SankeyDiagramNode[], settings: SankeyDiagramSettings) {
-        nodes.forEach((node: SankeyDiagramNode) => {
-            let nodeSettings: SankeyDiagramNodePositionSetting = this.getNodeSettings(node.label.name, settings);
-            node.settings = nodeSettings;
-        });
+
+        function getNode(nodeName: string, settings) {
```
Comment (uve) [inline src/sankeydiagram.ts]: add interface to settings

File: src/sankeydiagram.ts [hunk 7]
```diff
@@ -978,20 +1040,90 @@ export class SankeyDiagram implements IVisual {
         ];
     }
 
+
+    private deserializeLinks(links: string): SankeyDiagramLink[] {
+        let serializedLinks: string[] = JSON.parse(links);
+        let deserializedNodes: SankeyDiagramNode[] = [];
+        let deserializedLinks = serializedLinks.map(stringLink => {
```
Comment (uve) [inline src/sankeydiagram.ts]: add interface to stringLink

File: src/sankeydiagram.ts [hunk 8]
```diff
@@ -1318,14 +1442,54 @@ export class SankeyDiagram implements IVisual {
...
+                .sort((a, b) => {
+                    let x, y;
+                    if (sortBy === "name") {
+                        x = a.label.name;
+                        y = b.label.name;
+                    } else if (sortBy === "weight") {
+                        x = Math.max(a.inputWeight, a.outputWeight);
+                        y = Math.max(b.inputWeight, b.outputWeight);
+                    }
+                    return ascending ? ((x < y) ? -1 : ((x > y) ? 1 : 0)) : -((x < y) ? -1 : ((x > y) ? 1 : 0));
```
Comment (uve) [inline src/sankeydiagram.ts]: replace it by:
(ascending ? 1: -1) * ((x < y) ? -1 : ((x > y) ? 1 : 0))

File: .vscode/settings.json [hunk 9]
```diff
@@ -8,7 +8,6 @@
     ".tmp": true
   },
   "files.exclude": {
-    ".tmp": true
```
Comment (uve) [inline .vscode/settings.json]: do not remove this line

File: capabilities.json [hunk 10]
```diff
@@ -249,10 +262,31 @@
                     "type": {
                         "text": true
                     }
+                },
+                "linkPositions": {
+                    "displayName": "Link positions",
+                    "displayNameKey": "Visual_LinkPositions",
```
Comment (uve) [inline capabilities.json]: When you create a new **displayNameKey** entry, don't forget to add English string with the same key here:
https://github.com/microsoft/powerbi-visuals-sankey/blob/master/stringResources/en-US/resources.resjson#L30

File: capabilities.json [hunk 11]
```diff
@@ -249,10 +262,31 @@
...
+                    "type": {
+                        "text": true
+                    }
+                },
+                "columns": {
+                    "displayName": "Columns",
+                    "displayNameKey": "Visual_Columns",
```
Comment (uve) [inline capabilities.json]: Add a new key to en-US/resources.resjson as well

File: capabilities.json [hunk 12]
```diff
@@ -249,10 +262,31 @@
...
+                    "type": {
+                        "text": true
+                    }
+                },
+                "appState": {
+                    "displayName": "Application State",
+                    "displayNameKey": "Visual_AppState",
```
Comment (uve) [inline capabilities.json]: Add a new key to en-US/resources.resjson as well

File: src/sankeydiagram.ts (hunk 5 above)
Comment (MulyukovAidar) [reply src/sankeydiagram.ts]: Not the same. _linkPositions always keeps all link data of an application, links keep data that is required for rendering. If filter was applied _linkPositions will still have all the data parsed from appState, but links will only contain a subset - the links that should be rendered.

**Pull request closed without merging**